RFCs: Lay out rebalancing plans/designs for the 1.1 release

a-robinson opened the pull request.

a-robinson committed:
RFCs: Lay out rebalancing plans/designs for the 1.1 release

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16296)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj5j1Mj1dH2t76eL49:-Klj5j1Mj1dH2t76eL4A:b-824x5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L89)):*
> ```Markdown
> the average value for simplicity, but computing a few percentiles may be more
> valuable in helping pick out the true outliers that rebalancing would have the
> greatest effect on the cluster’s balance.
> ```

FYI, we already periodically loop over the replicas on a store and compute stats. Computing percentiles for replica sizes/load would be straightforward to add.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj5qdhHGEcqzh8vh90:-Klj5qdhHGEcqzh8vh91:b-ny6ihg) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L91)):*
> ```Markdown
> greatest effect on the cluster’s balance.
> 
> ## Bytes-based rebalancing
> ```

I've been calling this size-based rebalancing and have a mild preference for that term. 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj7z4vqkwgT4E5oQta:-Klj7z4vqkwgT4E5oQtb:b-pbe3d9) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L118)):*
> ```Markdown
> change the replica logic as part of this work to take disk size into account,
> such that we’ll balance replicas per GB of disk rather than absolute number of
> replicas.
> ```

Something to be aware of here is that the current replica-count balancing causes rebalancing to occur after a split. With size-based balancing a split would not cause rebalancing. This is probably desirable, but it is something to pay attention to while these heuristics are developed. 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj8O17stvYTFn1f4mV:-Klj8O17stvYTFn1f4mW:b-u5vu8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L131)):*
> ```Markdown
> replicas are). As a result, I think we’ll want to start tracking the number of
> applied commands on each replica instead of relying on the existing leaseholder
> QPS.
> ```

I think there is a need to flesh out out load will be combined with size to make rebalancing decisions. There is an issue somewhere that discussions "momentum"-based rebalancing. That is, combining load (velocity) with size (mass) and balancing the momentum of all replicas across the cluster. Perhaps you have other ideas here.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6Vh1L9aoz2o3-kj8:-Klj6Vh1L9aoz2o3-kj9:bh3kmii) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L182)):*
> ```Markdown
> streaming top-k problem, and there are plenty of algorithms that have been
> written about that should work for us given that we only need approximate
> results.
> ```

Tracking the top-k keys seems like an application of existing algorithms. But how to track the top-k key spans seems novel. Am I missing something here about how this is a straightforward extension of existing top-k algorithms?

---

*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-Klj6q2RKwQp9dsfww-9:b8d9f9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
> ```Markdown
>    other ranges), start recording the approximate top-k key spans.
>    Correspondingly, if a range's QPS drops down and we had been tracking its
>    top-k key spans, we should notice this and stop.
> ```

An alternative to tracking top-k key spans would be to select the split point that would normally be chosen (e.g. the mid-key) and start tracking QPS separately on both sides of the split point. If there is a reasonably even amount of traffic to both sides of the split point (e.g. 70/30 or 80/20), then perform the split. This would be much less precise than tracking the top-k key spans, but seems more feasible (as I'm not quite sure how the tracking of the top-k key spans would work). This could probably be generalized to choosing N equidistant possible split points.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-:-Klj8piRTdq9YFYlxNwQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 98 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KlsU0spts-gdIL9Gzfe:-KlsU0spts-gdIL9Gzff:b-8ro8c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L98)):*
> ```Markdown
> 1. We already gossip each store’s total disk capacity and unused disk capacity.
>    We could start trying to balance unused disk capacity across all the nodes of
>    the cluster. That would mean that in the case of heterogenous disk sizes,
> ```

I think you want heterogeneous, not heterogenous.

(sorry 😬)

---

*[docs/RFCS/rebalancing-for-1-1.md, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KlsTvn94HogNt1X3sFx:-KlsTvn94HogNt1X3sFy:b-ubftd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L204)):*
> ```Markdown
> 5. Sit back and let load-based rebalancing do its thing.
> 
> This will might take a bit of work to finish, and isn’t critical for 1.1, but
> ```

"will might"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-Klsg0G_1KcBYXizMhmc:b-713v4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

An alternative to tracking top-k key spans would be to select the split point that would normally be chosen (e.g. the mid-key) and start tracking QPS separately on both sides of the split point. If there is a reasonably even amount of traffic to both sides of the split point (e.g. 70/30 or 80/20), then perform the split. This would be much less precise than tracking the top-k key spans, but seems more feasible (as I'm not quite sure how the tracking of the top-k key spans would work). This could probably be generalized to choosing N equidistant possible split points.
</blockquote></details>

How about binary search the split point until traffic is roughly even on both sides?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj5j1Mj1dH2t76eL49:-KlsceTv3lVrxnDaJoWX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L89)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

FYI, we already periodically loop over the replicas on a store and compute stats. Computing percentiles for replica sizes/load would be straightforward to add.
</blockquote></details>

Done.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 91 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj5qdhHGEcqzh8vh90:-KlschRU4pLS4zdsb2rp:b-1eh3wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L91)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I've been calling this size-based rebalancing and have a mild preference for that term. 
</blockquote></details>

Sure, done.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 98 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KlsU0spts-gdIL9Gzfe:-Klsd1S-zWR2ppUDvkJc:b-uiivmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L98)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think you want heterogeneous, not heterogenous.

(sorry 😬)
</blockquote></details>

Heh, I didn't know that heterogenous was anything more than a misspelling before now. Done.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj7z4vqkwgT4E5oQta:-KlsehZl5Z_CCGJBpvrX:bqbhcjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L118)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Something to be aware of here is that the current replica-count balancing causes rebalancing to occur after a split. With size-based balancing a split would not cause rebalancing. This is probably desirable, but it is something to pay attention to while these heuristics are developed. 
</blockquote></details>

Yup, worth keeping an eye on, but it's probably good not to move it if the QPS or count-based rebalancing don't find it to be worth moving.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj8O17stvYTFn1f4mV:-KlseyGWa9ByB_sPcb4y:b-m4vnxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L131)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think there is a need to flesh out out load will be combined with size to make rebalancing decisions. There is an issue somewhere that discussions "momentum"-based rebalancing. That is, combining load (velocity) with size (mass) and balancing the momentum of all replicas across the cluster. Perhaps you have other ideas here.
</blockquote></details>

You've mentioned this a few times now in different places, so I think it might be best if you flesh out what you're thinking in a little more detail. Are you suggesting that we should be careful about rebalancing replicas that have a higher momentum (i.e. load and size) because moving them is potentially more disruptive?

---

*[docs/RFCS/rebalancing-for-1-1.md, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6Vh1L9aoz2o3-kj8:-KlslUN8-fVwW2cRaBDH:b-qpmil3) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L182)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Tracking the top-k keys seems like an application of existing algorithms. But how to track the top-k key spans seems novel. Am I missing something here about how this is a straightforward extension of existing top-k algorithms?
</blockquote></details>

Well, depending on how precise we're trying to be, it seems like there's a pretty direct mapping. Just take the span header from the request to the replica and use it directly, with `(*Span).Equal` serving as the comparator that enables the top-k algorithm to treat it as a single value. This won't be as precise about which keys are touched as, say, our timestamp cache interval tree, but it'll give us a good idea of the most common requests to the range.

The logic for finding a split point after the data has been gathered can then deal with reasoning about spans that are overlapping.

Am I missing something? Is there a reason that won't work?

---

*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-Klsn2qx_Db8EVz18weU:bdnzb53) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

How about binary search the split point until traffic is roughly even on both sides?
</blockquote></details>

This is definitely an interesting idea if I'm missing something about why the top-k approach isn't feasible.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KlsTvn94HogNt1X3sFx:-KlsdHrJ0kKK3Jy8a8i7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L204)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

"will might"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

benesch commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj8O17stvYTFn1f4mV:-KlsusW8LkIdJ8dyWxQL:b4kl5au) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L131)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

You've mentioned this a few times now in different places, so I think it might be best if you flesh out what you're thinking in a little more detail. Are you suggesting that we should be careful about rebalancing replicas that have a higher momentum (i.e. load and size) because moving them is potentially more disruptive?
</blockquote></details>

I'm just asking for more detail on how we'll be balancing in multiple dimensions. The "balancing based on multiple factors" section references some of the difficulties here, but I'm still missing specifics on how the heuristics will utilize the multiple factors. Combining the factors into a single weight is one approach. Or there could be ad-hoc heuristics to switch between which factor to prioritize. 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6Vh1L9aoz2o3-kj8:-KlsvbXlZzmFo5uHj1f9:bck6iw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L182)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Well, depending on how precise we're trying to be, it seems like there's a pretty direct mapping. Just take the span header from the request to the replica and use it directly, with `(*Span).Equal` serving as the comparator that enables the top-k algorithm to treat it as a single value. This won't be as precise about which keys are touched as, say, our timestamp cache interval tree, but it'll give us a good idea of the most common requests to the range.

The logic for finding a split point after the data has been gathered can then deal with reasoning about spans that are overlapping.

Am I missing something? Is there a reason that won't work?
</blockquote></details>

Ok, this extra detail was useful. How are you going to select the split point given the top-k hot keys? I think there is a straightforward answer here about choosing the key from the top-k hot keys which provides the best load balance. Worthwhile to call this out in this doc.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-KlsxRQeO0DqOgnF7GYQ:bmbyh20) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

This is definitely an interesting idea if I'm missing something about why the top-k approach isn't feasible.
</blockquote></details>

I'm aware of streaming top-k algorithms, but not their implementation. Are there any concerns about performance of adding such an algorithm to the core code path? The binary search approach that I alluded to and Arjun fleshed out would involve adding a single comparison to the core code path which could make it more palatable. Something to keep in mind as this is developed (perhaps add to an "alternatives" section).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-KltGa1lPL_TO5Rb6K8W:b-tkfvev) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm aware of streaming top-k algorithms, but not their implementation. Are there any concerns about performance of adding such an algorithm to the core code path? The binary search approach that I alluded to and Arjun fleshed out would involve adding a single comparison to the core code path which could make it more palatable. Something to keep in mind as this is developed (perhaps add to an "alternatives" section).
</blockquote></details>

How much more memory will this take up for each range?  All this metadata can start to add up quickly.

If we are able to automatically merge small ranges back together, we can do more blind splitting (just split the range in half) and not worry about dealing with the consequences as much since they will be cleaned up automatically.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 12 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltEDRbj3QPooFitHG2:-KltEDRcvI1DyAZe4vZW:b-iv5ns7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L12)):*
> ```Markdown
> # Summary
> 
> Lay out plans for which rebalancing improvements to make in the 1.1 release and
> ```

Since this is aiming for 1.1 and 1.2, perhaps change the title and summary to reflect that?

This summary is wrong.  
Describe some rebalancing improvements and investigate which ones are the most tractable and helpful in the 1.1 and 1.2 release timeframe.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 159 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltEiL-ty4NOsW9o83U:-KltEiL-ty4NOsW9o83V:b-h1k0k2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L159)):*
> ```Markdown
> ## Load-based splitting
> 
> Load-based splitting is conceptually pretty simple, but will likely some edge
> ```

likely `produce` some edge 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 173 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltEwPdwrW_B_n66d3G:-KltEwPdwrW_B_n66d3H:b-70lmer) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L173)):*
> ```Markdown
> 4. A range gets a lot of scan requests or other requests that touch multiple
>    keys. Splitting could actually make things worse by flipping an operation
>    from a single-range operation into a multi-range one.
> ```

I think there can be a danger here if we don't handle the reverse case of merging low load small ranges back together.

For example:
Let's say there's a high load on a table that tapers off over time that had a significant number of inserts.  Each new row is very hot for the first day, then for the next day that load is halved, then the next day it's halved again, etc.  
This would potentially lave a huge number of tiny replicas that should probably be merged back together after the load has dropped significantly.  Or else all the gains we received by splitting can be negated by having a massive number of tiny ranges.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltHMpWirzIgPcWTOK_:-KltHMpWirzIgPcWTOKa:b-hkzkqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L206)):*
> ```Markdown
> would be a nice addition and comes with much less downside risk than something
> like load-based replica locality. We’ll try to get to it if we have the time,
> otherwise can implement it for 1.2.
> ```

So what is actually going to be implemented.  I'd like to see a conclusion on what will actually be worked on for 1.1 and 1.2.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-KltOOBWb1ofi7QJZKhN:bu0r1r4) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
> select the split point that would normally be chosen (e.g. the mid-key)

The "normal" process for choosing a split point doesn't aim for the midpoint, it aims for the key with a distance from the start key of `max_range_size / 2`. So it's already going to be a slight tweak to the process to find the midpoint of a range that hasn't yet reached its max size. 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltMTFANBz17UlnG2ox:-KltMTFANBz17UlnG2oy:b-yi66dn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L42)):*
> ```Markdown
> load is on that range and on each of the candidate nodes.
> 
> ## Moving replicas closer to where their load is coming from ("load-based replica locality")
> ```

This one seems lower-priority than the others to me, since it only really applies when there are more localities than the replication factor. 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltMblkU_Q83of7rbZY:-KltMblkU_Q83of7rbZZ:b-1athac) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L63)):*
> ```Markdown
> load across the cluster by rebalancing one of the halves. 
> 
> This is so important for performance that we already support manually
> ```

As a data point, bigtable had load-based splitting, but (at the time I was using it), it was still important to pre-split (and scatter) before launches. 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 166 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltR0EAfJy24TwAnguw:-KltR0EAfJy24TwAngux:b-1faene) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L166)):*
> ```Markdown
> 
> 2. A range gets a lot of requests for just a couple of individual keys.
>    Splitting will help if and only if the split is between the hot keys.
> ```

If the split is between the hot keys *and* the hot keys are hot *in different queries*. If the same query often touches both hot keys then it's probably better to leave them together (similar to point 4).

---

*[docs/RFCS/rebalancing-for-1-1.md, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltPjXSA0RQhL8Gn552:-KltPjXTTfq5-Tzi0Pj9:b-7mlebk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L168)):*
> ```Markdown
>    Splitting will help if and only if the split is between the hot keys.
> 
> 3. A range gets a lot of requests for just a single key. Splitting won’t help at
> ```

I wonder if there is any way to collect data about this before we implement it. My hunch is that ranges are unlikely to have multiple keys that are independently "hot" after an initial "big bang" period when the table is growing from nothing. 

Instead of tracking per-key stats, I wonder if we might get most of the benefit by splitting more aggressively (i.e. at some fraction of max_range_size) until the number of ranges per table is at least as large as the number of nodes in the cluster.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-:-KltRK33l5EBG9xWLHzf:bb6ywmx)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6Vh1L9aoz2o3-kj8:-KltSKZXaRsmnIaYHudb:b-5h4kbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L182)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok, this extra detail was useful. How are you going to select the split point given the top-k hot keys? I think there is a straightforward answer here about choosing the key from the top-k hot keys which provides the best load balance. Worthwhile to call this out in this doc.
</blockquote></details>

Added a basic workable approach to the steps below.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-KltFbeIt-kvQf8_6HhD:b-cjmpdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L194)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

How much more memory will this take up for each range?  All this metadata can start to add up quickly.

If we are able to automatically merge small ranges back together, we can do more blind splitting (just split the range in half) and not worry about dealing with the consequences as much since they will be cleaned up automatically.
</blockquote></details>

There are some pretty efficient algorithms out there (e.g. the space-saving algorithm in http://www.cse.ust.hk/~raywong/comp5331/References/EfficientComputationOfFrequentAndTop-kElementsInDataStreams.pdf offers O(1) updates while using Θ(k) memory and being quite simple to implement. http://dimacs.rutgers.edu/~graham/pubs/papers/freq.pdf has some good comparisons of different algorithms, and indicates that in 2008 they were able to hit greater than 10k updates per millisecond with some of the algorithms.

So I'm not too concerned about adding one of them to the core code path, but any of them would be more expensive and complex than the simulation-based approach. I've added an alternatives section -- we can compare the options head-to-head when we get to the point of implementing it.

@bramgruneir: It'll take up a single pointer's worth of memory for most ranges, since if they aren't high-traffic we'll just nil out the relevant structure. For the couple highest traffic ranges while we're investigating them, it'll take O(k) memory to track the hottest k spans. Tracking something like 100 spans for each of the hottest couple ranges every once in a while is unlikely to be a big deal.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 12 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltEDRbj3QPooFitHG2:-KltNfs0cQL_GZiX0WCX:bemmlsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L12)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Since this is aiming for 1.1 and 1.2, perhaps change the title and summary to reflect that?

This summary is wrong.  
Describe some rebalancing improvements and investigate which ones are the most tractable and helpful in the 1.1 and 1.2 release timeframe.
</blockquote></details>

Is this good enough? 

---

*[docs/RFCS/rebalancing-for-1-1.md, line 159 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltEiL-ty4NOsW9o83U:-KltJ0AndFRGVeFWGweQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L159)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

likely `produce` some edge 
</blockquote></details>

Done.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 173 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltEwPdwrW_B_n66d3G:-KltJACljAqe5H_gqLn9:bxe98er) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L173)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I think there can be a danger here if we don't handle the reverse case of merging low load small ranges back together.

For example:
Let's say there's a high load on a table that tapers off over time that had a significant number of inserts.  Each new row is very hot for the first day, then for the next day that load is halved, then the next day it's halved again, etc.  
This would potentially lave a huge number of tiny replicas that should probably be merged back together after the load has dropped significantly.  Or else all the gains we received by splitting can be negated by having a massive number of tiny ranges.
</blockquote></details>

Yup, good point. Added a drawbacks section mentioning that.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltHMpWirzIgPcWTOK_:-KltIhvif3ENc9CWTrzb:b8oy9lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L206)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

So what is actually going to be implemented.  I'd like to see a conclusion on what will actually be worked on for 1.1 and 1.2.
</blockquote></details>

What isn't clear from the subsections?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj8O17stvYTFn1f4mV:-KltaGV5eI6lQl-vFfRW:b-7u0kki) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing-for-1-1.md#L131)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm just asking for more detail on how we'll be balancing in multiple dimensions. The "balancing based on multiple factors" section references some of the difficulties here, but I'm still missing specifics on how the heuristics will utilize the multiple factors. Combining the factors into a single weight is one approach. Or there could be ad-hoc heuristics to switch between which factor to prioritize. 
</blockquote></details>

I've added a brief explanation but it's admittedly pretty hand-wavy. Basically, prioritize balancing the dimensions that are least balanced by using the replicas that are the strongest outliers in those dimensions.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltMTFANBz17UlnG2ox:-KltTaYPHamG88v8aKyv:blo06z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This one seems lower-priority than the others to me, since it only really applies when there are more localities than the replication factor. 
</blockquote></details>

Yup, agreed. It's definitely not happening for 1.1, and likely not for 1.2 either. I included it to explicitly de-prioritize it.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltMblkU_Q83of7rbZY:-KltTi_qo-B5vVZnUNZS:brvmop0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L63)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

As a data point, bigtable had load-based splitting, but (at the time I was using it), it was still important to pre-split (and scatter) before launches. 
</blockquote></details>

Yup, and that'll most likely be true here as well for really quick spikes.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 166 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltR0EAfJy24TwAnguw:-KltXDvamjHOe9tibeFK:brx9836) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L166)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If the split is between the hot keys *and* the hot keys are hot *in different queries*. If the same query often touches both hot keys then it's probably better to leave them together (similar to point 4).
</blockquote></details>

Worded this a little more carefully now, thanks.

---

*[docs/RFCS/rebalancing-for-1-1.md, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltPjXSA0RQhL8Gn552:-KltXuQEDV3XMZiQN3vY:buuruuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing-for-1-1.md#L168)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I wonder if there is any way to collect data about this before we implement it. My hunch is that ranges are unlikely to have multiple keys that are independently "hot" after an initial "big bang" period when the table is growing from nothing. 

Instead of tracking per-key stats, I wonder if we might get most of the benefit by splitting more aggressively (i.e. at some fraction of max_range_size) until the number of ranges per table is at least as large as the number of nodes in the cluster.
</blockquote></details>

We can't realistically collect data about this before 1.1 without having a bunch of representative workload generators. If we're fine pushing off the actual splitting until 1.2, we could put some of this tracking logic into the next release and report anonymous statistics on hot key distributions, but if we've already implemented the stats-tracking then adding on the splitting isn't much extra work.

I'm not sure about your latter suggestion. It'll help for workloads where you're quickly growing a table from nothing, but seemingly not for much else.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltPjXSA0RQhL8Gn552:-KltmVrsRj2yU2g7D2Qk:b-8b9lcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing_for_1_1.md#L168)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

We can't realistically collect data about this before 1.1 without having a bunch of representative workload generators. If we're fine pushing off the actual splitting until 1.2, we could put some of this tracking logic into the next release and report anonymous statistics on hot key distributions, but if we've already implemented the stats-tracking then adding on the splitting isn't much extra work.

I'm not sure about your latter suggestion. It'll help for workloads where you're quickly growing a table from nothing, but seemingly not for much else.
</blockquote></details>

Yeah. I'm making the claim that the main case in which load-based splitting would help is actually the "big bang" scenario and that it's not worth implementing the rest of this proposal. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-Klu2N93KD5LhpiwROb6:b-z3kziu) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing_for_1_1.md#L194)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

There are some pretty efficient algorithms out there (e.g. the space-saving algorithm in http://www.cse.ust.hk/~raywong/comp5331/References/EfficientComputationOfFrequentAndTop-kElementsInDataStreams.pdf offers O(1) updates while using Θ(k) memory and being quite simple to implement. http://dimacs.rutgers.edu/~graham/pubs/papers/freq.pdf has some good comparisons of different algorithms, and indicates that in 2008 they were able to hit greater than 10k updates per millisecond with some of the algorithms.

So I'm not too concerned about adding one of them to the core code path, but any of them would be more expensive and complex than the simulation-based approach. I've added an alternatives section -- we can compare the options head-to-head when we get to the point of implementing it.

@bramgruneir: It'll take up a single pointer's worth of memory for most ranges, since if they aren't high-traffic we'll just nil out the relevant structure. For the couple highest traffic ranges while we're investigating them, it'll take O(k) memory to track the hottest k spans. Tracking something like 100 spans for each of the hottest couple ranges every once in a while is unlikely to be a big deal.
</blockquote></details>

My concern is more if we implement the binary search approach and store usage metrics per key.  Just something to keep an eye on is all.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltHMpWirzIgPcWTOK_:-KltrzNKpPkddLTao2sL:bqyfpf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing_for_1_1.md#L206)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

What isn't clear from the subsections?
</blockquote></details>

I just think a conclusion section at the end or a conclusion subsection in each section on what will be implemented and what wont would be very helpful.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-KltsGmsZRPvuZAFxk0r:b21b2a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
> ```Markdown
> weight we give to that dimension. These heuristics will allow us to combine
> the different dimensions into a single final score, and should be covered by a
> large number of test cases to ensure stability in different scenarios.
> ```

This heuristic is clearly going to be troublesome.  My intuition is telling me that it's going to be a mix of special cases (for all edge cases) and some coefficient of weights for each feature.    I'm not sure of the best approach to determine those weights, my guess is hand tuning at first, but I think in the end, it will probably require a simulator using a collection of different setups.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 194 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klj6q2RKwQp9dsfww-8:-KlycSMu9OVGnUjnhtQL:b-r38yqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/790d9ac58166107a68c448cca1571d1754219d39/docs/RFCS/rebalancing_for_1_1.md#L194)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

My concern is more if we implement the binary search approach and store usage metrics per key.  Just something to keep an eye on is all.
</blockquote></details>

The binary search approach will only operate on the top couple ranges by QPS and will seemingly just need to store

1. The possible split key
2. The count of requests to the left of the key, to the right of the key, or that touch both sides of the key.

We'll probably keep around the counts for the last few split keys attempted, but I don't see anything worrying.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 168 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltPjXSA0RQhL8Gn552:-KlycKpU0lNic3Ecln1O:bxhfcdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing_for_1_1.md#L168)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah. I'm making the claim that the main case in which load-based splitting would help is actually the "big bang" scenario and that it's not worth implementing the rest of this proposal. 
</blockquote></details>

I see, that's a fair point. Added to the alternatives section.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltHMpWirzIgPcWTOK_:-KlyctxQ5atLVWX8dK2g:bxmdf5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/dad65d187c24224af9134d67bec7abb9221951fd/docs/RFCS/rebalancing_for_1_1.md#L206)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I just think a conclusion section at the end or a conclusion subsection in each section on what will be implemented and what wont would be very helpful.
</blockquote></details>

Ack, I'm sorry if it isn't super clear, but don't want to spend lots of time wordsmithing the RFC at this point when I could get to work on the code instead. The gist is that everything in the Design section is targeted for 1.1 unless stated otherwise (i.e. load-based locality is pushed back until there's demand, load-based splitting is best effort for 1.1 but first on the chopping block). I'll try to do better next time!

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-KlydrVkTb6UPWDD3k-0:b-pqsqxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This heuristic is clearly going to be troublesome.  My intuition is telling me that it's going to be a mix of special cases (for all edge cases) and some coefficient of weights for each feature.    I'm not sure of the best approach to determine those weights, my guess is hand tuning at first, but I think in the end, it will probably require a simulator using a collection of different setups.
</blockquote></details>

Yeah, as written this seems pretty flaky. Do you have any thoughts on how to make the combining of factors more solid?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-KlyjqoW981kbmBbU9ed:bzd5m75) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Yeah, as written this seems pretty flaky. Do you have any thoughts on how to make the combining of factors more solid?
</blockquote></details>

The alternative I've been suggesting (if we only have size and load factors) is to determine a replica's weight as `size * load` and then balance that metric across the nodes in the cluster. In reality, this might be `size * log(load)` or something like that. Undoubtedly, experimentation will be needed.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 95 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klykr7PK83Ci-Du9bZT:-Klykr7PK83Ci-Du9bZU:b-k3hn80) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdef56e4de47878c894d39f2132171605ba64ad0/docs/RFCS/rebalancing_for_1_1.md#L95)):*
> ```Markdown
> overfull for that dimension toward stores that are less full for the dimension
> (and vice versa for low percentiles and underful stores under the expectation
> that the removed replicas can be replaced by higher percentile replicas). the
> ```

`s/. the/. The/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-KlymudfzdaXyN1n6Eyl:b3rm80j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The alternative I've been suggesting (if we only have size and load factors) is to determine a replica's weight as `size * load` and then balance that metric across the nodes in the cluster. In reality, this might be `size * log(load)` or something like that. Undoubtedly, experimentation will be needed.
</blockquote></details>

Could you explain the intuition behind balancing the product of size and load?

---

*[docs/RFCS/rebalancing_for_1_1.md, line 95 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-Klykr7PK83Ci-Du9bZT:-Klymo5FcgsP4e8_jcwm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdef56e4de47878c894d39f2132171605ba64ad0/docs/RFCS/rebalancing_for_1_1.md#L95)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/. the/. The/g`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-Klyo5p4EQ_3IR_tZ5JT:b-h2wzr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Could you explain the intuition behind balancing the product of size and load?
</blockquote></details>

A range with more load is considered "heavier" and we then balance a single "weight" metric across the cluster. I can imagine some drawbacks to this approach, though. Space might not be balanced well at all. There is also a certain amount of background load associated with the size of a range. For example, background consistency checks, sending Raft snapshots, etc. We want to avoid too much space imbalance as it can exacerbate the impact of a node outage. We want to maximize load balancing in order to get the most utilization out of the cluster. I'm rambling a bit here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-KlysPn2XnTlqH20Ulj6:bu39yz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

A range with more load is considered "heavier" and we then balance a single "weight" metric across the cluster. I can imagine some drawbacks to this approach, though. Space might not be balanced well at all. There is also a certain amount of background load associated with the size of a range. For example, background consistency checks, sending Raft snapshots, etc. We want to avoid too much space imbalance as it can exacerbate the impact of a node outage. We want to maximize load balancing in order to get the most utilization out of the cluster. I'm rambling a bit here. 
</blockquote></details>

You've stated a couple of the downsides, but I'm not understanding the benefit of using this combined momentum metric rather than considering disk usage and load as independent variables. Is it just to reduce the number of factors in the decision?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-Klz0eMxnGJkxDaqTiP9:b-gm8opi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

You've stated a couple of the downsides, but I'm not understanding the benefit of using this combined momentum metric rather than considering disk usage and load as independent variables. Is it just to reduce the number of factors in the decision?
</blockquote></details>

How do you combine the factors to arrive at a decision? We need each node to be making the same decisions in order to avoid thrashing. It might be a failure of my imagination, but it seems this results in either a single score associated with each range to balance across the cluster, or something equivalent. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-KlzvUHEQEKuA3ETHJ9u:b-j2bvfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How do you combine the factors to arrive at a decision? We need each node to be making the same decisions in order to avoid thrashing. It might be a failure of my imagination, but it seems this results in either a single score associated with each range to balance across the cluster, or something equivalent. 
</blockquote></details>

I think going with the independent factor route is going to be easier to tune.  That combined with some hard limits.
(w<sub>1</sub>X<sub>1</sub> + w<sub>2</sub>X<sub>2</sub> + w<sub>3</sub>X<sub>3</sub> + w<sub>4</sub>X<sub>4</sub> ) /4 is going to be much easier to tune by hand than anything that combines them directly.

And then add in hard maxes and mins.

If we set the weights as cluster settings, it would be possible to have a program try to tune and active cluster, which would be really cool.  For all we know, using bytes and replica counts might be redundant and only one of them is needed (or maybe it's the opposite).

Furthermore, I can see different weights being used for different cluster sizes.  a size 4 to 6 node cluster might have different needs than a 32 to 64 node cluster.

There's a lot more exciting stuff to do here, I'm going to see if I can find some literature on it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
I'm going to throw this into its final comment period and get to work on it, but will wait until you think it's good to go, @BramGruneir.

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-Km2ddayo_HYpkBaXwgn:b4sjbuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I think going with the independent factor route is going to be easier to tune.  That combined with some hard limits.
(w<sub>1</sub>X<sub>1</sub> + w<sub>2</sub>X<sub>2</sub> + w<sub>3</sub>X<sub>3</sub> + w<sub>4</sub>X<sub>4</sub> ) /4 is going to be much easier to tune by hand than anything that combines them directly.

And then add in hard maxes and mins.

If we set the weights as cluster settings, it would be possible to have a program try to tune and active cluster, which would be really cool.  For all we know, using bytes and replica counts might be redundant and only one of them is needed (or maybe it's the opposite).

Furthermore, I can see different weights being used for different cluster sizes.  a size 4 to 6 node cluster might have different needs than a 32 to 64 node cluster.

There's a lot more exciting stuff to do here, I'm going to see if I can find some literature on it.


</blockquote></details>

Finding some literature would be great. I expect that for much of anything we do here we're going to need larger imbalance tolerances than we currently have (we have a 5% tolerance for replica count imbalance). We're also going from using a factor that is constant (i.e. each replica = 1) to numbers that can quickly change over time (primarily QPS, but even disk usage can grow over time).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-:-Km2ePveotZtR6sg2GKf:b-8k36he)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/rebalancing_for_1_1.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-KltsGmsZRPvuZAFxk0q:-Km2euaoxDj5m3dV7s2q:b-e6v1mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9acc204f579f92d1cfdc9065ddf3261d3d22e9c/docs/RFCS/rebalancing_for_1_1.md#L99)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Finding some literature would be great. I expect that for much of anything we do here we're going to need larger imbalance tolerances than we currently have (we have a 5% tolerance for replica count imbalance). We're also going from using a factor that is constant (i.e. each replica = 1) to numbers that can quickly change over time (primarily QPS, but even disk usage can grow over time).
</blockquote></details>

I agree, that fuzzy buffer zone is probably going to have to be wider.  But regardless, this is going to be fun!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16296#-:-Km2eswwowI6NVe_rMAf:bnfp4nl)*
<!-- Sent from Reviewable.io -->
